### DISCUSSION TITLE
Can I pass an object to a model before it's validate method is called?

### DISCUSSION THREAD

**Discussion by mjackson-mim** [playframework/playframework#12828] [Q&A]
I'm using Play! 3.0.2 with Java 21 on the back end, and I have a model that implements [Constraints.Validatable\<T\>](https://www.playframework.com/documentation/2.8.x/api/java/index.html), that I'm binding to a form using [Form\<T\>.bindFromRequest(request)](https://www.playframework.com/documentation/2.8.x/api/java/index.html). I need to use a service class in `validate()`'s body, but I want to pass in the implementation of the service class at runtime. Is there a way to do one of the following:

1. Use Guice to inject the implementation into the model's constructor when it's created.
2. Get access to the model's getters before the form is validated.
3. Something else I'm not thinking of :)

An example:
```java
public class MyModel implements Constraints.Validatable<List<ValidationError>> {

	private String url;

	private UrlValidator urlValidator;

	@Inject
	public MyModel(UrlValidator urlValidator) {                   // Thought 1
		this.urlValidator = urlValidator;
	}

	public void setUrlValidator(UrlValidator urlValidator) {      // Thought 2
		this.urlValidator = urlValidator;
	}

	public List<ValidationError> validate() {
		List<ValidationError> errors = new LinkedList<>();

		if (!urlValidator.isValid(url)) {
			errors.add(new ValidationError("url", "URL is not valid."));
		}
		return errors;
	}
}
```

```java
public MyController extends FormHandlingController {

	public Result myEndpoint(Http.Request request) {
		Form<T> form = factory.form(clazz).bindFromRequest(request); // Throws error
		if (form.hasErrors()) {
			return ok();
		} else {
			return badRequest();
		}
	}
}
```

Things I've tried / work-arounds I know about

- I've tried calling the setters on the result of `Form.get()`, with no luck.
- I know I could move the validation logic to the controller, but I would prefer the code to be in the validate method if possible.
- I know I could call `Guice.createInjector().getInstance(UrlValidator.class)`, but that is [bad practice](https://github.com/google/guice/wiki/InjectingTheInjector).

Any help is greatly appreciated!

**Comment by GithubUser8080** [comment]
Maybe it would be better and more conformant to the usual Jakarta validation if you added a custom constraint instead of injecting the validation in the model.

`
@ValidUrl
private String url;
`

ValidUrl can be your own constraint annotation, that uses 
`
validatedBy = UrlValidator.class`

Then whenever your model is validated, a UrlValidator will be automatically injected and will try to validate the field without you polluting the model.
